frappe/core/api/user_invitation.py

```diff
@@ -0,0 +1,149 @@
+import frappe
+from frappe import _
+from frappe.model.document import Document
+
+
+def _validate_app_name(app_name: str) -> None:
+	if app_name not in frappe.get_installed_apps():
+		frappe.throw(_("Application is not installed"))
+
+
+def _get_allowed_roles(app_name: str) -> list[str]:
+	user_invitation_hook = frappe.get_hooks("user_invitation", app_name=app_name)
+	allowed_roles = user_invitation_hook.get("allowed_invite_roles")
+	return allowed_roles if allowed_roles is not None else []
+
+
+def _validate_role(role: str, allowed_roles: list[str]) -> None:
+	if role not in allowed_roles:
+		frappe.throw(_("Role is not in the allowed roles list"))
+
+
+def _validate_emails(emails: str) -> list[str]:
+	frappe.utils.validate_email_address(emails, throw=True)
+	email_list = frappe.utils.split_emails(emails)
+	if not email_list:
+		frappe.throw(_("No email addresses found"))
+	return email_list
+
+
+def _get_existing_user_emails(email_list: list[str]) -> list[str]:
+	return frappe.db.get_all("User", filters={"email": ["in", email_list]}, pluck="email")
+
+
+def _get_existing_invited_emails(email_list: list[str], allowed_roles: list[str] | None) -> list[str]:
+	filters = {"email": ["in", email_list], "status": "Pending"}
+	if allowed_roles is not None:
+		filters["role"] = ["in", allowed_roles]
+	return frappe.db.get_all("User Invitation", filters=filters, pluck="email")
+
+
+@frappe.whitelist()
+def invite_by_email(
+	emails: str, role: str, redirect_to_path: str, app_name: str | None = None
+) -> dict[str, list[str]]:
+	frappe.only_for(["System Manager", "User Invitation Manager"])
+	app_name = app_name if app_name is not None else "frappe"
+	_validate_app_name(app_name)
+	allowed_roles: list[str] = []
+	if app_name != "frappe":
+		allowed_roles = _get_allowed_roles(app_name)
+		_validate_role(role, allowed_roles)
+	email_list = _validate_emails(emails)
+	existing_user_emails = _get_existing_user_emails(email_list)
```

Comment:
For 1/2 line utils, if they're not used more than once, prefer inlining them. It's easier to read then jumping across functions.